# 1) Understand the difference between vertical tables and horizontal tables.

In the database world, tables are used to store data, so there are two different ways to structure these tables, among which are the Horizontal and Vertical structures. The horizontal structure is generally  based on the partitioning of the data in columns i.e. each column contains only part of the data. This kind of table is generally more used due to its easy of handling when compared to other structures of tables, but in some contexts i.e. deploying an app and then realize that it needs more data that are not in the table turning necessary adding a new column to tha dataframe might cause some problems and maybe will be necessary to re-structure thw whole deploy again which will be quite laborious to handle.

On the other hand, the vertical tables have the data and the variables structured in rows, which turns it easy to add new data into the dataframe making it a better option in contexts which are dynamic and work with a huge data volumn, so the kind of table that you'll use to work is highly dependent of the context of your goal, so it is recomended to check carefully what is the goal of your project? How dynamic is it? And based on these question use the kind of data structure that better fits on your project. In the classification context, Xie et al. (2013) have found that representations in horizontal  can robustly improve the performance.

## Reference

Xie, J; Xu, B; Chuang, Z. Horizontal and Vertical Ensemble with Deep Representation for
Classification. Presented at the ICML Workshop on Representation Learning, Atlanta, Georgia, USA, 2013.: https://arxiv.org/pdf/1306.2759.pdf



# 2) Find examples of vertical and horizontal tables

## About the data:

The data we're going to use in this exercise was scrapped from the site: "https://amazon.com" which is a platmform dedicated to sell a gigantic variety of products. Among the products comercialized in Amazon, we've choosen to work with the lis of mangás (japanese comics) found in the website. This dataset is structured in four columns which are: 

### Title
The title of the Mangá that we collected the data

### Price
The price of the mangá

### Rating
That is the rating of the mangá, if it has a good note in a scale from 0 to 5 stars

### URL
This is the URL of the product, where you can find to buy.

The code used to scrap the data may be found in the following link: https://github.com/Viniciusfcfranca/Web_Scrapping_Polifact/blob/main/Web%20Scrapping%20Amazon.ipynb

In [1]:
#Importing the necessary packages
import numpy as np
import pandas as pd
df = pd.read_csv("prices.csv")
#Showing the dataframe
df.head(10)

Title    Price  \
0                                   Jujutsu Kaisen 0    $8.99   
1                                      The Two Lions   $12.59   
2                            Kaiju No. 8, Vol. 1 (1)    $9.99   
3                  nichijou 15th anniversary box set   $82.78   
4  Dragon Ball Complete Box Set: Vols. 1-16 with ...  $116.19   
5  Demon Slayer: Kimetsu no Yaiba--Stories of Wat...    $7.49   
6                         Jujutsu Kaisen, Vol. 1 (1)    $9.76   
7                  Vagabond, Vol. 1 (VIZBIG Edition)   $17.99   
8  Solo Leveling, Vol. 1 (comic) (Solo Leveling (...   $16.06   
9                           Spy x Family, Vol. 1 (1)    $9.99   

               Rating                                                URL  
0  4.8 out of 5 stars  https://www.amazon.com/Jujutsu-Kaisen-0-Gege-A...  
1  4.7 out of 5 stars  https://www.amazon.com/Two-Lions-Nagisa-Furuya...  
2  4.9 out of 5 stars  https://www.amazon.com/Kaiju-No-Vol-Naoya-Mats...  
3  4.9 out of 5 stars  https://www.amazon.com/nichijou-15th-anniversa...  
4  4.8 out of 5 stars  https://www.amazon.com/Dragon-Ball-Complete-Bo...  
5  4.7 out of 5 stars  https://www.amazon.com/Demon-Slayer-Kimetsu-Ya...  
6  4.9 out of 5 stars  https://www.amazon.com/Jujutsu-Kaisen-Vol-1/dp...  
7  4.8 out of 5 stars  https://www.amazon.com/Vagabond-Vol-VIZBIG-Tak...  
8  4.8 out of 5 stars  https://www.amazon.com/Solo-Leveling-Vol-comic...  
9  4.9 out of 5 stars  https://www.amazon.com/Spy-Family-Vol-Tatsuya-...

In the context of this exercise, we'll work only with the columns title, price and rating so, we'll drop the other column, but before this, we'll clean the price column to it be classified as a numeric variable.

In [2]:
df['Price'] = df['Price'].map(lambda x: x.lstrip('$').rstrip('aAbBcC'))
df['Price']= df['Price'].astype(float) 

In [3]:
df

Title   Price  \
0                                     Jujutsu Kaisen 0    8.99   
1                                        The Two Lions   12.59   
2                              Kaiju No. 8, Vol. 1 (1)    9.99   
3                    nichijou 15th anniversary box set   82.78   
4    Dragon Ball Complete Box Set: Vols. 1-16 with ...  116.19   
..                                                 ...     ...   
303                 Fist of the North Star, Vol. 1 (1)   17.99   
304                            One Piece, Vol. 98 (98)    9.49   
305  I'm the Villainess, So I'm Taming the Final Bo...    8.99   
306                                       FLCL Omnibus   17.99   
307                           Ero Ninja Scrolls Vol. 2   13.99   

                 Rating                                                URL  
0    4.8 out of 5 stars  https://www.amazon.com/Jujutsu-Kaisen-0-Gege-A...  
1    4.7 out of 5 stars  https://www.amazon.com/Two-Lions-Nagisa-Furuya...  
2    4.9 out of 5 stars  https://www.amazon.com/Kaiju-No-Vol-Naoya-Mats...  
3    4.9 out of 5 stars  https://www.amazon.com/nichijou-15th-anniversa...  
4    4.8 out of 5 stars  https://www.amazon.com/Dragon-Ball-Complete-Bo...  
..                  ...                                                ...  
303  4.9 out of 5 stars  https://www.amazon.com/Fist-North-Star-Vol-1/d...  
304  4.9 out of 5 stars  https://www.amazon.com/One-Piece-Vol-98/dp/197...  
305  4.7 out of 5 stars  https://www.amazon.com/Villainess-Taming-Final...  
306  4.8 out of 5 stars  https://www.amazon.com/FLCL-Omnibus-Gainax/dp/...  
307                      https://www.amazon.com/Ero-Ninja-Scrolls-Vol-2...  

[308 rows x 4 columns]

In [4]:
df.dtypes

Title      object
Price     float64
Rating     object
URL        object
dtype: object

In [5]:
df = df.drop(['URL'], 1)

In [6]:
df.head()

Title   Price  \
0                                   Jujutsu Kaisen 0    8.99   
1                                      The Two Lions   12.59   
2                            Kaiju No. 8, Vol. 1 (1)    9.99   
3                  nichijou 15th anniversary box set   82.78   
4  Dragon Ball Complete Box Set: Vols. 1-16 with ...  116.19   

               Rating  
0  4.8 out of 5 stars  
1  4.7 out of 5 stars  
2  4.9 out of 5 stars  
3  4.9 out of 5 stars  
4  4.8 out of 5 stars

This dataframe is an example of horizontal dataframe where each column represents part of the total data, spliting it in columns. Now let's see an example of how would be a vertical dataframe.

In [7]:
df2 = df.set_index('Title').stack().reset_index(name='values') #using the function stack in the original dataframe
df2 = pd.DataFrame(df2) #converting to dataframe
df2

Title level_1  \
0                                     Jujutsu Kaisen 0   Price   
1                                     Jujutsu Kaisen 0  Rating   
2                                        The Two Lions   Price   
3                                        The Two Lions  Rating   
4                              Kaiju No. 8, Vol. 1 (1)   Price   
..                                                 ...     ...   
611  I'm the Villainess, So I'm Taming the Final Bo...  Rating   
612                                       FLCL Omnibus   Price   
613                                       FLCL Omnibus  Rating   
614                           Ero Ninja Scrolls Vol. 2   Price   
615                           Ero Ninja Scrolls Vol. 2  Rating   

                 values  
0                  8.99  
1    4.8 out of 5 stars  
2                 12.59  
3    4.7 out of 5 stars  
4                  9.99  
..                  ...  
611  4.7 out of 5 stars  
612               17.99  
613  4.8 out of 5 stars  
614               13.99  
615                      

[616 rows x 3 columns]

In [8]:
df2 = df2.rename(columns={'level_1': 'variable'})
df2 = df2[['variable','Title','values']]

In [9]:
df2

variable                                              Title  \
0      Price                                   Jujutsu Kaisen 0   
1     Rating                                   Jujutsu Kaisen 0   
2      Price                                      The Two Lions   
3     Rating                                      The Two Lions   
4      Price                            Kaiju No. 8, Vol. 1 (1)   
..       ...                                                ...   
611   Rating  I'm the Villainess, So I'm Taming the Final Bo...   
612    Price                                       FLCL Omnibus   
613   Rating                                       FLCL Omnibus   
614    Price                           Ero Ninja Scrolls Vol. 2   
615   Rating                           Ero Ninja Scrolls Vol. 2   

                 values  
0                  8.99  
1    4.8 out of 5 stars  
2                 12.59  
3    4.7 out of 5 stars  
4                  9.99  
..                  ...  
611  4.7 out of 5 stars  
612               17.99  
613  4.8 out of 5 stars  
614               13.99  
615                      

[616 rows x 3 columns]

This dataframe is an example of a vertical dataframe because the variables are in rows being able to include new variables and values withoud the necessity to insert new columns.

# 3) Choose a categorical variable in the vertical basis and use the 'transpose' funtion in the 'Pandas' package to transpose the table and create a contigency table.

In [10]:
df2_transposed = df2.set_index('variable').transpose()
df2_transposed

variable             Price              Rating          Price  \
Title     Jujutsu Kaisen 0    Jujutsu Kaisen 0  The Two Lions   
values                8.99  4.8 out of 5 stars          12.59   

variable              Rating                    Price  \
Title          The Two Lions  Kaiju No. 8, Vol. 1 (1)   
values    4.7 out of 5 stars                     9.99   

variable                   Rating                              Price  \
Title     Kaiju No. 8, Vol. 1 (1)  nichijou 15th anniversary box set   
values         4.9 out of 5 stars                              82.78   

variable                             Rating  \
Title     nichijou 15th anniversary box set   
values                   4.9 out of 5 stars   

variable                                              Price  \
Title     Dragon Ball Complete Box Set: Vols. 1-16 with ...   
values                                               116.19   

variable                                             Rating  ...  \
Title     Dragon Ball Complete Box Set: Vols. 1-16 with ...  ...   
values                                   4.8 out of 5 stars  ...   

variable                               Price  \
Title     Fist of the North Star, Vol. 1 (1)   
values                                 17.99   

variable                              Rating                    Price  \
Title     Fist of the North Star, Vol. 1 (1)  One Piece, Vol. 98 (98)   
values                    4.9 out of 5 stars                     9.49   

variable                   Rating  \
Title     One Piece, Vol. 98 (98)   
values         4.9 out of 5 stars   

variable                                              Price  \
Title     I'm the Villainess, So I'm Taming the Final Bo...   
values                                                 8.99   

variable                                             Rating         Price  \
Title     I'm the Villainess, So I'm Taming the Final Bo...  FLCL Omnibus   
values                                   4.7 out of 5 stars         17.99   

variable              Rating                     Price  \
Title           FLCL Omnibus  Ero Ninja Scrolls Vol. 2   
values    4.8 out of 5 stars                     13.99   

variable                    Rating  
Title     Ero Ninja Scrolls Vol. 2  
values                              

[2 rows x 616 columns]

# 4) Understand the concept of contingency and incidence matrix.

## Contingency Matrix
In statistics, a contingency table (also known as a cross tabulation or crosstab) is a type of table in a matrix format that displays the (multivariate) frequency distribution of the variables. They are heavily used in survey research, business intelligence, engineering, and scientific research. They provide a basic picture of the interrelation between two variables and can help find interactions between them. summarizes the conditional frequencies of two attributes and shows how these two attributes are dependent on each other with the information on a partition of universe generated by these attributes. The contingency term was firstly used by Karl Pearson in 1904. 

## Reference

Karl Pearson, F.R.S. (1904). Mathematical contributions to the theory of evolution. Dulau and Co.: https://archive.org/details/cu31924003064833.

Shusaku Tsumoto and Shoji Hirano, "Contingency matrix theory," 2007 IEEE International Conference on Systems, Man and Cybernetics, 2007, pp. 3778-3783, doi: 10.1109/ICSMC.2007.4413917.
https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=4413917

## Example of Contingency Matrix.



| Sex/Handness | Right-handed | Left-handed | Total |
|--- |--- |--- |--- |
| Male | 43 | 9 | 52 | 
| Female | 44 | 4 | 48 | 
| Total | 87 | 13 | 100 | 

## Incidence Matrix
In mathematics, an incidence matrix is a logical matrix that shows the relationship between two classes of objects, usually called an incidence relation with a huge applicability in the graphs theorem. The incidence matrix A of an undirected graph has a row for each vertex and a column for each edge of the graph whereas The incidence matrix A of a directed graph has a row for each vertex and a column for each edge of the graph.


## Reference
Fred E. Szabo. The Linear Algebra Survival Guide (2015), Academic Press, doi: https://doi.org/10.1016/C2012-0-06836-6.


## Example of Incidence Matrix

\begin{bmatrix} 1,1,1,0 \\ 1,0,0,0 \\ 0,1,0,1 \\ 0,0,1,1 \end{bmatrix}

## This matrix represents the graph bellow:

![](graph_incidence.jpeg)


# 5) Study the function melt and apply in the data.

## Melt

The melt function is useful to massage a DataFrame into a format where one or more columns are identified as variables (id_vars), while all other columns, considered measured variables (value_vars), leaving just two non-identifier columns, ‘variable’ and ‘value’.

## Reference
Pandas Documentation: https://pandas.pydata.org/docs/reference/api/pandas.melt.html

In [11]:
df.melt(id_vars=['Title'])

Title variable  \
0                                     Jujutsu Kaisen 0    Price   
1                                        The Two Lions    Price   
2                              Kaiju No. 8, Vol. 1 (1)    Price   
3                    nichijou 15th anniversary box set    Price   
4    Dragon Ball Complete Box Set: Vols. 1-16 with ...    Price   
..                                                 ...      ...   
611                 Fist of the North Star, Vol. 1 (1)   Rating   
612                            One Piece, Vol. 98 (98)   Rating   
613  I'm the Villainess, So I'm Taming the Final Bo...   Rating   
614                                       FLCL Omnibus   Rating   
615                           Ero Ninja Scrolls Vol. 2   Rating   

                  value  
0                  8.99  
1                 12.59  
2                  9.99  
3                 82.78  
4                116.19  
..                  ...  
611  4.9 out of 5 stars  
612  4.9 out of 5 stars  
613  4.7 out of 5 stars  
614  4.8 out of 5 stars  
615                      

[616 rows x 3 columns]

# 6) Use the transpose function in SQL to achieve the same result as the Pandas transpose.

### At first, let's turn this dataframe into a SQL table, so at first we're going to convert it into an excell file and then import to the MySQL

In [12]:
df2.to_excel('prices.xlsx')

Right now we're going to import it into a sql schema.

Into the mysql program, we select one of the schemas we have, then we'd clicked in tables section and after this we had clicked with the right bottom of the mouse and clicked in 'create new table'. It openned a new window in te sql to we create the table and name the columns. We'd named the table as 'fact_checker' and named the columns with the same names of the columns in the dataframe then created applied to create the table.

After this, we clicked with the right bottom in the new table we had created and choose the option "table data import wizard" to import the file to the mysql and after this, our table was filled with data.

After that I used the following querry in Mysql:

SELECT *

FROM (SELECT Variable as Variable,

Title as Title,

'Values' as 'Values' 

FROM puthon_activities.prices

group by Title, Variable) em_linha, PIVOT (SUM(valor) for Variable in (['Price'],['Rating']) em_colunas

ORDER BY 1

Which returned to me:



| variables | Price | Rating | Price | Rating | Price | ... | Price | Rating |
|--- |--- |--- |--- |--- |--- |--- |--- |--- |
| Title| Jujutsu Kaisen 0 | Jujutsu Kaisen 0 | The Two Lions | The Two Lions |Kaiju No. 8, Vol. 1 (1) | ... | Ero Ninja Scrolls Vol. 2 |Ero Ninja Scrolls Vol. 2 | 
| Values | 8.99 | 4.8 out of 5 stars | 12.59 | 4.7 out of 5 stars |9.99 | ...| 13.99 |    | 
